File: seqspec/seqspec_split.py
"""Split module for seqspec.

This module provides functionality to split seqspec files into one file per modality.
"""
from pathlib import Path
from argparse import ArgumentParser, RawTextHelpFormatter, Namespace
from typing import List

from seqspec.utils import load_spec
from seqspec.Assay import Assay


def setup_split_args(parser) -> ArgumentParser:
    """Create and configure the split command subparser."""
    subparser = parser.add_parser(
        "split",
        description="""
Split seqspec file into one file per modality.

Examples:
seqspec split -o split spec.yaml  # Split spec into modalities
---
""",
        help="Split seqspec file by modality",
        formatter_class=RawTextHelpFormatter,
    )
    subparser_required = subparser.add_argument_group("required arguments")

    subparser.add_argument("yaml", help="Sequencing specification yaml file", type=str)
    subparser_required.add_argument(
        "-o",
        "--output",
        metavar="OUT",
        help="Path to output files",
        type=Path,
        required=True,
    )
    return subparser


def validate_split_args(parser: ArgumentParser, args: Namespace) -> None:
    """Validate the split command arguments."""
    if not Path(args.yaml).exists():
        parser.error(f"Input file does not exist: {args.yaml}")

    if Path(args.output).exists() and Path(args.output).is_file():
        parser.error(f"Output path exists: {args.output}")


def run_split(parser: ArgumentParser, args: Namespace) -> None:
    """Run the split command."""
    validate_split_args(parser, args)

    spec = load_spec(args.yaml)
    specs = seqspec_split(spec)

    prefix = "spec." if args.output.name == "" else f"{args.output.name}."
    for spec_m in specs:
        modality = spec_m.list_modalities()[0]
        output_path = args.output / f"{prefix}{modality}.yaml"
        spec_m.to_YAML(output_path)


def seqspec_split(spec: Assay) -> List[Assay]:
    """Split spec into one file per modality.

    Args:
        spec: The Assay object to split

    Returns:
        List of Assay objects, each containing a single modality
    """
    specs = []
    modalities = spec.list_modalities()

    # Make a new spec per modality
    for modality in modalities:
        info = {
            "assay_id": spec.assay_id,
            "name": spec.name,
            "doi": spec.doi,
            "date": spec.date,
            "description": spec.description,
            "modalities": [modality],
            "lib_struct": spec.lib_struct,
            "library_kit": spec.library_kit,
            "library_protocol": spec.library_protocol,
            "sequence_kit": spec.sequence_kit,
            "sequence_protocol": spec.sequence_protocol,
            "sequence_spec": spec.get_seqspec(modality),
            "library_spec": [spec.get_libspec(modality)],
            "seqspec_version": spec.seqspec_version,
        }
        spec_m = Assay(**info)
        spec_m.update_spec()
        specs.append(spec_m)

    return specs

File: src/seqspec_split.rs
use crate::models::assay::Assay;
use crate::utils;
use clap::Args;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Args)]
pub struct SplitArgs {
    #[clap(help = "Sequencing specification yaml file", required = true)]
    yaml: PathBuf,

    #[clap(
        short,
        long,
        help = "Path to output files",
        value_name = "OUT",
        required = true
    )]
    output: PathBuf,
}

fn validate_split_args(args: &SplitArgs) {
    if !args.yaml.exists() {
        eprintln!("Please use `seqspec split -h` for help.");
        std::process::exit(1);
    }
    if args.output.exists() && args.output.is_file() {
        eprintln!("Output path exists: {}", args.output.display());
        std::process::exit(1);
    }
}

pub fn run_split(args: &SplitArgs) {
    validate_split_args(args);
    let spec = utils::load_spec(&args.yaml);
    let specs = seqspec_split(&spec);

    // Determine prefix (name + "." or "spec.")
    let prefix = args
        .output
        .file_name()
        .and_then(|s| s.to_str())
        .filter(|s| !s.is_empty())
        .map(|s| format!("{}.", s))
        .unwrap_or_else(|| "spec.".to_string());

    // Ensure output directory exists
    let _ = fs::create_dir_all(&args.output);

    for mut spec_m in specs {
        let modality = spec_m.list_modalities().get(0).cloned().unwrap_or_default();
        let out_path = args.output.join(format!("{}{}.yaml", prefix, modality));
        spec_m.update_spec();
        let bytes = spec_m.to_bytes().expect("serialize to YAML");
        let mut f = fs::File::create(out_path).expect("create output file");
        f.write_all(&bytes).expect("write YAML");
    }
}

pub fn seqspec_split(spec: &Assay) -> Vec<Assay> {
    let mut specs: Vec<Assay> = Vec::new();
    let modalities = spec.list_modalities();
    for modality in modalities {
        let sequence_spec = spec.get_seqspec(&modality);
        let library_spec = vec![spec
            .get_libspec(&modality)
            .expect("modality not found in library_spec")];

        let mut spec_m = Assay::new(
            spec.assay_id.clone(),
            spec.name.clone(),
            spec.doi.clone(),
            spec.date.clone(),
            spec.description.clone(),
            vec![modality.clone()],
            spec.lib_struct.clone(),
            sequence_spec,
            library_spec,
            spec.sequence_protocol.clone(),
            spec.sequence_kit.clone(),
            spec.library_protocol.clone(),
            spec.library_kit.clone(),
            spec.seqspec_version.clone(),
        );
        spec_m.update_spec();
        specs.push(spec_m);
    }
    specs
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::load_spec;

    fn dogma_spec() -> Assay {
        load_spec(&PathBuf::from("tests/fixtures/spec.yaml"))
    }

    #[test]
    fn test_split_produces_one_per_modality() {
        let spec = dogma_spec();
        let mods = spec.list_modalities();
        let splits = seqspec_split(&spec);
        assert_eq!(splits.len(), mods.len());
    }

    #[test]
    fn test_split_preserves_metadata() {
        let spec = dogma_spec();
        let splits = seqspec_split(&spec);
        for s in &splits {
            assert_eq!(s.assay_id, spec.assay_id);
            assert_eq!(s.name, spec.name);
            assert_eq!(s.doi, spec.doi);
        }
    }

    #[test]
    fn test_split_single_modality_each() {
        let spec = dogma_spec();
        let splits = seqspec_split(&spec);
        for s in &splits {
            assert_eq!(s.modalities.len(), 1);
            assert_eq!(s.library_spec.len(), 1);
        }
    }

    #[test]
    fn test_split_rna_reads_only() {
        let spec = dogma_spec();
        let splits = seqspec_split(&spec);
        let rna_spec = splits.iter().find(|s| s.modalities[0] == "rna").unwrap();
        for r in &rna_spec.sequence_spec {
            assert_eq!(r.modality, "rna");
        }
    }

    #[test]
    fn test_split_atac_reads_only() {
        let spec = dogma_spec();
        let splits = seqspec_split(&spec);
        let atac_spec = splits.iter().find(|s| s.modalities[0] == "atac").unwrap();
        for r in &atac_spec.sequence_spec {
            assert_eq!(r.modality, "atac");
        }
    }

    #[test]
    fn test_split_preserves_total_reads() {
        let spec = dogma_spec();
        let total_reads = spec.sequence_spec.len();
        let splits = seqspec_split(&spec);
        let split_total: usize = splits.iter().map(|s| s.sequence_spec.len()).sum();
        assert_eq!(split_total, total_reads);
    }
}
